infor-design/enterprise
# 3767/3777/3845 - Multiselect Improvements

**Explain the _details_ for making this change. What existing problem does the pull request solve?**
This PR fixes a number of usability issues and bugs with the Multiselect (Dropdown) component:
- Multiselects that display tags and have a "Select All" checkbox will no longer render both the text and tags display when choosing "Select All". (#3767)
- Multiselects with many options that are also displaying tags can now be configured to constrain the tag list with a maximum height using the `tagListMaxHeight` setting, which also makes the tag display scrollable. (#3767)
- Multiselects that have a "Select All" checkbox enabled will not display this checkbox if there are no options present to choose from. (#3777)
- Previously there were different code paths for "Select All" behavior when clicking on the item versus using the keyboard to activate the option.  This has been resolved and it now uses the same logic for both operations. (n/a)
- When a filter term is present, by default, Multiselects will only select all items in the current filter when choosing "Select All".  This is configurable and can be put back to the original behavior using the `selectAllFilterOnly` setting. (#3845)

**Related github/jira issue (required)**:
Closes #3767 
Closes #3777 
Closes #3845 

**Steps necessary to review your pull request (required)**:
Pull this branch, build, run the demoapp.  Then:

_Test rendering behavior (#3767):_ @byannalam
- Open http://localhost:4000/components/multiselect/test-select-all.html to see text-only rendering
- Open the Multiselect and choose "Select All"
- Close the Multiselect
- when the list updates, notice it only renders the text-based results.  No tags should be present.
- Open http://localhost:4000/components/multiselect/test-select-all-tags.html
- Open the Multiselect and choose "Select All"
- The Multiselect list should update immediately, and should only contain results in the form of a Tag List.

_Test JS errors when using keyboard selection (n/a):_
- Open http://localhost:4000/components/multiselect/test-select-all.html
- Open the Multiselect and use the keyboard arrow keys to navigate to "Select All".  Press enter to activate it.
- Close the Multiselect to commit the results.
- Open a developer tools console and ensure there are no JS errors.

_Test empty Multselect behavior with Select All enabled (#3777):_ @pwpatton
- Open http://localhost:4000/components/multiselect/test-select-all-no-opts.html
- Open the Multiselect
- The Multiselect should be empty, and it should NOT contain a "Select All" button.

_Test improved Filter + Select All logic (#3845):_ @hohnzy @thomas-blackwell
- Open http://localhost:4000/components/multiselect/test-select-all-tags.html
- Open the Multiselect and click "Select All".  Every item in the list should become selected.
- Deselect "Select All".  All items should become deselected.
- Type "New" into the filter.
- Click "Select All".  Only the four items displayed in the filter should become selected.

**Included in this Pull Request**:
- [x] e2e or functional tests for the bugs or features.
- [x] notes to the change log.
